man/crashlistout.1
.TH CRASHLISTOUT 1 2014-03-27 "Johan Billing" "CrashMail"
.SH NAME
crashlistout \- List the mail files waiting to go out in the specified 
directory.
.SH SYNOPSIS
.B crashlistout
DIRECTORY [outbounddir] ZONE [zonenumber] PATTERN [string] [VERBOSE]
.SH DESCRIPTION
Checks the specified directory for outgoing mail files and prints the information
that was found. If no outbound directory is specified, it uses the path specified
in the environment variable CMOUTBOUND. If a zone is not specified, it uses that
defined the environment variable CMOUTBOUNDZONE or defaults to 2.

.SH OPTIONS
.TP
.I (DIRECTORY) <string>
The directory being checked.
.TP
.I (ZONE) <string>
Specify the primary zone. Defaults to 2.
.TP
.I (PATTERN) <string>
The specific node pattern to listed.
.TP
.I VERBOSE
If the listing should be verbose.
.SH "SEE ALSO"
.BR crashmail (1),
.\".SH FILES
.\".SH BUGS
.SH AUTHOR
CrashMail is written by Johan Billing <billing@df.lth.se>.
.PP
This manual page was written by Robert James Clay <jame@rocasa.us>.


ChangeLog
2014-05-16 19:12  Robert James Clay  <jame@rocasa.us>

  - Merge changes by Niklas Lindholm from the merge of them that
    Lars Kellogg-Stedman did in his repository.

  - Correct two instances of mis-spelling 'noticeable' as 'noticable' in the
    ChangeLog file.

  - Correct an instance of mis-spelling 'messages' as 'messsages' in the
    ChangeLog file.


2016-09-02 01:34  Niklas Lindholm <niklas@nilin.se>

 - Adding extra debug logging

 - Changing so that HighWater (1.msg) is updated every time an area is scanned
   and not just when a local message is exported.

 - Tweaking the logging so the message about updating hwm is only logged when
   the hwm is actually updated.


2014-05-01 03:34  Robert James Clay  <jame@rocasa.us>

 - Release as version 1.6

 - Add missing man page for the crashlistout binary.

 - Add thanks to Lars Kellogg-Stedman to the README file.

 - Remove duplicate entry from ChangeLog dated '2013-02-12 11:18'.


2014-03-06 16:45  Robert James Clay  <jame@rocasa.us>

 - Release as version 1.5


2014-03-06 14:14  Lars Kellogg-Stedman  <lars@seas.harvard.edu>

 - Merge fix of type mismatches and warnings through the code.

 - Merge using off_t for seek and tell operations.

 - Merge adding missing commands to man page.


2014-03-06 13:49  Robert James Clay  <jame@rocasa.us>

 - Add a text file with message testing content.

 - Correct name of crashwrite option for msg text in tests/02-crashwrite-test.sh.


2014-03-06 08:15  Lars Kellogg-Stedman  <lars@seas.harvard.edu>

 - Merge conversion from ulong/ushort to stdint types in src/cmnllib/, src/jamlib/,
   src/crashmail/, src/shared/, and src/tools/.


2013-02-17 19:21  Robert James Clay  <jame@rocasa.us>

 - Release as version 1.4

 - README, doc/ReadMe.txt: lars@oddbit.com is the email address preferred by
   Lars Kellogg-Stedman.


2013-02-17 18:56  Lars Kellogg-Stedman  <lars@seas.harvard.edu>

 - tests/02-crashwrite.sh: added tests for named packets


2013-02-15 12:08  Robert James Clay  <jame@rocasa.us>

 - Release as version 1.3

 - Adapt tests/ directory to this version of Crashmail II.

 - Add copyright information for tests/roundup to README file.


2013-02-15 10:02  Lars Kellogg-Stedman  <lars@seas.harvard.edu>

 - tests/*: Merge in the addition of testing and a tests/ subdirectory done by
   Lars Kellogg-Stedman in his Crashmail port (github.com/larsks/crashmail.git).


2013-02-12 14:27  Robert James Clay  <jame@rocasa.us>

 - Release as version 1.2.1


2013-02-12 14:06  Lars Kellogg-Stedman  <lars@seas.harvard.edu>

 - src/crashmail/areafix.c: more pointer size fixes
 - src/tools/crashmaint.c: more pointer size fixes


2013-02-12 13:40  Robert James Clay  <jame@rocasa.us>

 - Release as version 1.2

 - src/tools/crashmaint.c: crashmaint has already had its version set as v1.2;
   with the release of Crashmail with an overall version of v1.2, synchronize
   its version with crashmail itself by setting it from src/crashmail/version.h
   instead of setting it directly in its code.


2013-02-12 12:03  Robert James Clay  <jame@rocasa.us>

 - Release as version 1.1.1


2013-02-12 12:27  Lars Kellogg-Stedman  <lars@seas.harvard.edu>

 - src/tools/crashwrite.c: Fix integer-length issues in crashwrite. Crashwrite
   declares pktnum as type ulong when it should really be uint32_t. Using "%lx"
   instead of "%x" to display this was resulting in invalid packet filenames.


2013-02-12 11:18  Robert James Clay  <jame@rocasa.us>

 - Release as version 1.1

 - src/tools/crashexport.c: crashexport has already had its version set as v1.1;
   with the release of Crashmail with an overall version of v1.1, synchronize
   its version with crashmail itself by setting it from src/crashmail/version.h
   instead of setting it directly in its code.


2013-02-11  Robert James Clay  <jame@rocasa.us>

 - Release as version 1.0

 - Move, rename, and update doc/History.txt to be the top level ChangeLog file.

 - The tools crashexport, crashgetnode, crashlist, crashlistout, and crashstats
   have already had their versions set as v1.0; with the release of Crashmail
   v1.0, synchronize their versions with crashmail itself by setting their
   versions from src/crashmail/version.h instead of setting it directly in
   their code.


2013-02-10 20:14  Lars Kellogg-Stedman  <lars@seas.harvard.edu>

 - src/tools/crashstats.c: Fixed 64-bit problems in crashstats, which made
   assumptions about the size of pointers that would cause it to fail on 64bit
   systems.


2013-02-10 17:30  Robert James Clay  <jame@rocasa.us>

 - Release version 0.76

 - Add a top level Makefile to the distribution.

 - Update and rename src/ReadMe.txt as the top level INSTALL file.

 - Derive new top level README file from the content of doc/ReadMe.txt.

 - Move the definition of the Crashmail version from src/crashmail/crashmail.h
   to the new file src/crashmail/version.h.


2013-02-10 12:55  Robert James Clay  <jame@rocasa.us>

 - Release version 0.75.1

 - src/crashmail/crashmail.h: Correct VERSION entries for v0.75 and add a
   definition for VERSION_PATCH.


2013-02-09  Robert James Clay  <jame@rocasa.us>

 - Release version 0.75

 - Update copyright information in doc/ReadMe.txt and convert the file to UTF-8.

 - Change the format of the doc/History.txt changelog entries since v0.71 to
   include date and author information as the header line, instead of just
   using a "Changes in version m.nn" line.


2013-01-21  Robert James Clay  <jame@rocasa.us>

 - Release version 0.74

 - Add usage of CPPFLAGS, CFLAGS, & LDFLAGS to the source files Makefile.linux,
   cmnllib/Makefile.linux, jamlib/Makefile.linux, and oslib_linux/Makefile.


2010-12-24  Robert James Clay  <jame@rocasa.us>

 - Release version 0.73


2010-08-10  Petr Salinger  <Petr.Salinger@seznam.cz>

 - Fix for a 'fails to build on GNU/kFreeBSD' bug on Debian (Bug#592481).


2009-09-19  Robert James Clay  <jame@rocasa.us>

 - Release version 0.72


2009-09-11  Peter Krefting  <peter@softwolves.pp.se>


src/tools/crashlistout.c
#include <string.h>
#include <stdio.h>
#include <stdlib.h>
#include <stdarg.h>
#include <ctype.h>

#include <shared/types.h>
#include <shared/mystrncpy.h>
#include <shared/path.h>
#include <shared/parseargs.h>
#include <shared/jblist.h>
#include <shared/node4d.h>

#include <oslib/os.h>
#include <oslib/osmem.h>
#include <oslib/osfile.h>
#include <oslib/osdir.h>
#include <oslib/osmisc.h>

#include <crashmail/version.h>

#define NET(x)  (x >> 16)
#define NODE(x) (x & 0xFFFF)

#define TYPE_CRASH    0
#define TYPE_DIRECT   1
#define TYPE_NORMAL   2
#define TYPE_HOLD     3
#define TYPE_REQUEST  4 

uchar *type_names[] = { "Crash", "Direct", "Normal", "Hold", "Request" };

ulong TotalFiles=0;
ulong TotalBytes=0;
ulong TotalRequests=0;

struct fileentry
{
   struct fileentry *Next;
   struct Node4D Node;
   uchar file[100];
   uchar dir[100];
   ulong size;
   time_t date;
   ulong type;
   bool flow;
};

struct Node4DPat
{
   uchar Zone[10];
   uchar Net[10];
   uchar Node[10];
   uchar Point[10];
};

uchar *cfg_Dir;
ulong cfg_Zone;
struct Node4DPat cfg_Pattern;
bool cfg_Verbose;

#define ARG_DIRECTORY  0
#define ARG_ZONE       1
#define ARG_PATTERN    2
#define ARG_VERBOSE    3

struct argument args[] =
   { { ARGTYPE_STRING, "DIRECTORY", ARGFLAG_AUTO,  NULL },
     { ARGTYPE_STRING, "ZONE",      ARGFLAG_AUTO,  NULL },
     { ARGTYPE_STRING, "PATTERN",   ARGFLAG_AUTO,  NULL },
     { ARGTYPE_BOOL,   "VERBOSE",   0,             NULL },
     { ARGTYPE_END,    NULL,        0,             0    } };

struct jbList list;

/* Some stuff for node pattern (taken from CrashMail) */

bool Parse4DPat(uchar *buf, struct Node4DPat *node)
{
   ulong c=0,tempc=0;
   uchar temp[10];
   bool GotZone=FALSE,GotNet=FALSE,GotNode=FALSE;

   strcpy(node->Zone,"*");
   strcpy(node->Net,"*");
   strcpy(node->Node,"*");
   strcpy(node->Point,"*");

   if(strcmp(buf,"*")==0)
      return(TRUE);

	for(c=0;c<strlen(buf);c++)
	{
		if(buf[c]==':')
		{
         if(GotZone || GotNet || GotNode) return(FALSE);
         strcpy(node->Zone,temp);
         GotZone=TRUE;
			tempc=0;
	   }
		else if(buf[c]=='/')
		{
         if(GotNet || GotNode) return(FALSE);
         strcpy(node->Net,temp);
         GotNet=TRUE;
			tempc=0;
		}
		else if(buf[c]=='.')
		{
         if(GotNode) return(FALSE);
         strcpy(node->Node,temp);
         node->Point[0]=0;
         GotNode=TRUE;
			tempc=0;
		}
		else if((buf[c]>='0' && buf[c]<='9') || buf[c]=='*' || buf[c]=='?')
		{
         if(tempc<9)
         {
            temp[tempc++]=buf[c];
            temp[tempc]=0;
         }
		}
		else return(FALSE);
	}

   if(GotZone && !GotNet)
   {
      strcpy(node->Net,temp);
   }
   else if(GotNode)
   {
      strcpy(node->Point,temp);
   }
   else
   {
      strcpy(node->Node,temp);
      strcpy(node->Point,"0");
   }

   return(TRUE);
}

int NodeCompare(uchar *pat,ushort num)
{
   uchar buf[10],c;
   sprintf(buf,"%u",num);

   if(pat[0]==0) return(0);

   for(c=0;c<strlen(pat);c++)
   {
      if(pat[c]=='*') return(0);
      if(pat[c]!=buf[c] && pat[c]!='?') return(1);
   }

   if(buf[c]!=0)
      return(1);

   return(0);
}

int Compare4DPat(struct Node4DPat *nodepat,struct Node4D *node)
{
   if(node->Zone!=0)
      if(NodeCompare(nodepat->Zone, node->Zone )!=0) return(1);

   if(NodeCompare(nodepat->Net,  node->Net  )!=0) return(1);
   if(NodeCompare(nodepat->Node, node->Node )!=0) return(1);
   if(NodeCompare(nodepat->Point,node->Point)!=0) return(1);

   return(0);
}

/* Some other functions from CrashMail */

char *unit(long i)
{
   static char buf[20];
   if ((i>10000000)||(i<-10000000)) sprintf(buf,"%luM",i/(1024*1024));
   else if ((i>10000)||(i<-10000)) sprintf(buf,"%luK",i/1024);
   else sprintf(buf,"%lu",i);
   return buf;
}

unsigned long hextodec(char *hex)
{
   char *hextab="0123456789abcdef";
   int c=0,c2=0;
   unsigned long result=0;

   for(;;)
   {
      for(c2=0;c2<16;c2++)
         if(tolower(hex[c]) == hextab[c2]) break;
		
		if(c2 == 16)
			return(result); /* End of correct hex number */

      result *= 16;
      result += c2;

      c++;
   }
}

void strip(char *str)
{
   int c;

   for(c=strlen(str)-1;str[c] < 33 && c>=0;c--) 
		str[c]=0;
}

/* Display entries in the list */

uchar *PrintNode(struct Node4D *node)
{
   static uchar buf[50];

   Print4D(node,buf);

   return(buf);
}

uchar *PrintDate(time_t date)
{
   static uchar buf[50];
   struct tm *tp;
   uchar *monthnames[]={"Jan","Feb","Mar","Apr","May","Jun","Jul","Aug","Sep","Oct","Nov","Dec","???"};

   tp=localtime(&date);

   sprintf(buf,"%02d-%s-%02d %02d:%02d",
      tp->tm_mday,
      monthnames[tp->tm_mon],
      tp->tm_year % 100,
      tp->tm_hour,
      tp->tm_min);

   return(buf);
}

uchar *PrintFlowSize(struct fileentry *fe)
{
   static uchar buf[50];
   uchar fullfile[200],line[200];
   osFile os;
   struct osFileEntry *osfe;
   ulong files,bytes;

   files=0;
   bytes=0;

   MakeFullPath(cfg_Dir,fe->file,fullfile,200);

   if(!(os=osOpen(fullfile,MODE_OLDFILE)))
   {
      sprintf(buf,"?/?");
      return(buf);
   }

   while(osFGets(os,line,200))
   {
      strip(line);

      if(line[0])
      {
         if(line[0] == '#' || line[0] == '^' || line[0] == '-')
            strcpy(line,&line[1]);
      
         if(stricmp(GetFilePart(line),line) == 0)
         {
            /* No path specified */

            MakeFullPath(fe->dir,line,fullfile,200);
            osfe=osGetFileEntry(fullfile);
         }
         else
         {
            osfe=osGetFileEntry(line);
         }

         if(osfe)
         {
            files++;
            bytes+=osfe->Size;

            TotalFiles++;
            TotalBytes+=osfe->Size;

            osFree(osfe);
         }
      }
   }

   osClose(os);

   sprintf(buf,"%s/%lu",unit(bytes),files);
   
   return(buf);
}

void DisplayFlowContents(struct fileentry *fe)
{
   uchar size[40],*todo;
   uchar fullfile[200],line[200];
   osFile os;
   struct osFileEntry *osfe;

   MakeFullPath(cfg_Dir,fe->file,fullfile,200);

   if(!(os=osOpen(fullfile,MODE_OLDFILE)))
   {
      printf("Failed to open file\n");
   }
   else
   {
      while(osFGets(os,line,200))
      {
         strip(line);

         if(line[0])
         {
            todo="";

            if(line[0] == '#' || line[0] == '^') todo="(To be truncated)";
            if(line[0] == '-')                   todo="(To be deleted)";

            if(line[0] == '#' || line[0] == '^' || line[0] == '-')
               strcpy(line,&line[1]);
      
            if(stricmp(GetFilePart(line),line) == 0)
            {
               /* No path specified */

               MakeFullPath(fe->dir,line,fullfile,200);
               osfe=osGetFileEntry(fullfile);
            }
            else
            {
               osfe=osGetFileEntry(line);
            }

            strcpy(size,"Not found");

            if(osfe)
            {
               sprintf(size,unit(osfe->Size));
               osFree(osfe);
            }

            printf(" %-39.39s %10s   %s\n",line,size,todo);
         }
      }

      osClose(os);
   }

   printf("\n");
}

uchar *PrintReqNums(struct fileentry *fe)
{
   static uchar buf[50];
   uchar fullfile[200],line[200];
   osFile os;
   ulong reqs;

   reqs=0;

   MakeFullPath(cfg_Dir,fe->file,fullfile,200);

   if(!(os=osOpen(fullfile,MODE_OLDFILE)))
   {
      sprintf(buf,"?/?");
      return(buf);
   }

   while(osFGets(os,line,200))
   {
      strip(line);

      if(line[0])
      {
         reqs++;
         TotalRequests++;
      }
   }

   sprintf(buf,"-/%lu",reqs);
   
   return(buf);
}

void DisplayReqContents(struct fileentry *fe)
{
   uchar fullfile[200],line[200];
   osFile os;

   MakeFullPath(cfg_Dir,fe->file,fullfile,200);

   if(!(os=osOpen(fullfile,MODE_OLDFILE)))
   {
      printf("Failed to open file\n");
   }
   else
   {
      while(osFGets(os,line,200))
      {
         strip(line);

         if(line[0])
            printf(" %s\n",line);
      }

      osClose(os);
   }

   printf("\n");
}

void DisplayFlow(struct fileentry *fe)
{
   printf("%-8.8s %-17.17s %-16.16s %7.7s   %s\n",
      type_names[fe->type],PrintNode(&fe->Node),PrintDate(fe->date),PrintFlowSize(fe),fe->file);

   if(cfg_Verbose)
      DisplayFlowContents(fe);
}

void DisplayPkt(struct fileentry *fe)
{
   uchar buf[100];

   sprintf(buf,"%s/1",unit(fe->size));

   printf("%-8.8s %-17.17s %-16.16s %7.7s   %s\n",
      type_names[fe->type],PrintNode(&fe->Node),PrintDate(fe->date),buf,fe->file);

   if(cfg_Verbose)
      printf("\n");
}

void DisplayReq(struct fileentry *fe)
{
   printf("%-8.8s %-17.17s %-16.16s %7.7s   %s\n",
      type_names[fe->type],PrintNode(&fe->Node),PrintDate(fe->date),PrintReqNums(fe),fe->file);

   if(cfg_Verbose)
      DisplayReqContents(fe);
}

int sortcompare(const void *f1, const void *f2)
{
   struct fileentry *e1,*e2;

   e1=*(struct fileentry **)f1;
   e2=*(struct fileentry **)f2;

   return Compare4D(&e1->Node,&e2->Node);
}

void sortlist(struct jbList *list)
{
   struct jbNode *jb,**buf,**work;
   ulong count=0;

   for(jb=list->First;jb;jb=jb->Next)
      count++;

   if(count < 2)
      return;

   if(!(buf=(struct jbNode **)osAlloc(count * sizeof(struct jbNode *))))
      return;

   work=buf;

   for(jb=list->First;jb;jb=jb->Next)
      *work++=jb;

   qsort(buf,(size_t)count,(size_t)sizeof(struct jbNode *),sortcompare);

   jbNewList(list);

   for(work=buf;count--;)
      jbAddNode(list,*work++);

   osFree(buf);
}

void addentry(uchar *dir,uchar *file,ulong type,struct Node4D *boss,bool flow)
{
   struct osFileEntry *fe;
   struct fileentry *entry;
   struct Node4D n4d;
   uchar buf[200];
   uchar buf2[200];
   ulong hex;

   hex=hextodec(file);

   if(boss)
   {
      Copy4D(&n4d,boss);
      n4d.Point = hex;
   }
   else
   {
      n4d.Zone = cfg_Zone;
      n4d.Net = NET(hex);
      n4d.Node = NODE(hex);
      n4d.Point = 0;
   }

   if(Compare4DPat(&cfg_Pattern,&n4d)!=0)
      return;

   if(dir) MakeFullPath(dir,file,buf,200);
   else    mystrncpy(buf,file,200);

   MakeFullPath(cfg_Dir,buf,buf2,200);

   if(!(fe=osGetFileEntry(buf2)))
   {
      return;
   }

   if(!(entry=osAlloc(sizeof(struct fileentry))))
   {
      osFree(fe);
      return;
   }

   Copy4D(&entry->Node,&n4d);

   if(dir)
   {
      MakeFullPath(dir,file,entry->file,100);
      MakeFullPath(cfg_Dir,dir,entry->dir,100);
   }
   else
   {
      mystrncpy(entry->file,file,100);
      mystrncpy(entry->dir,cfg_Dir,100);
   }

   mystrncpy(entry->file,buf,100);
   entry->size=fe->Size;
   entry->date=fe->Date;
   entry->type=type;
   entry->flow=flow;

   jbAddNode(&list,(struct jbNode *)entry);
   osFree(fe);
}

struct Node4D *scandir_boss;
uchar *scandir_dir;

void scandirfunc(uchar *file)
{
   uchar *extptr;
   ulong hex;

   if(strlen(file) != 12)
      return;

   extptr=&file[strlen(file)-4];

   if(stricmp(extptr,".PNT")==0 && !scandir_boss)
   {
      uchar buf[200];
      struct Node4D n4d;

      hex=hextodec(file);

      n4d.Zone = cfg_Zone;
      n4d.Net = NET(hex);
      n4d.Node = NODE(hex);
      n4d.Point = 0;

      scandir_dir = file;
      scandir_boss = &n4d;

      MakeFullPath(cfg_Dir,file,buf,200);
      osScanDir(buf,scandirfunc);

      scandir_dir = NULL;
      scandir_boss = NULL;
   }

   if(!stricmp(extptr,".REQ")) addentry(scandir_dir,file,TYPE_REQUEST,scandir_boss,TRUE);

   if(!stricmp(extptr,".CLO")) addentry(scandir_dir,file,TYPE_CRASH,scandir_boss,TRUE);
   if(!stricmp(extptr,".DLO")) addentry(scandir_dir,file,TYPE_DIRECT,scandir_boss,TRUE);
   if(!stricmp(extptr,".FLO")) addentry(scandir_dir,file,TYPE_NORMAL,scandir_boss,TRUE);
   if(!stricmp(extptr,".HLO")) addentry(scandir_dir,file,TYPE_HOLD,scandir_boss,TRUE);

   if(!stricmp(extptr,".CUT")) addentry(scandir_dir,file,TYPE_CRASH,scandir_boss,FALSE);
   if(!stricmp(extptr,".DUT")) addentry(scandir_dir,file,TYPE_DIRECT,scandir_boss,FALSE);
   if(!stricmp(extptr,".OUT")) addentry(scandir_dir,file,TYPE_NORMAL,scandir_boss,FALSE);
   if(!stricmp(extptr,".HUT")) addentry(scandir_dir,file,TYPE_HOLD,scandir_boss,FALSE);
}

int main(int argc, char **argv)
{
   uchar *var;
   struct fileentry *fe;

   if(!osInit())
      exit(OS_EXIT_ERROR);

   if(argc > 1 &&
	  (strcmp(argv[1],"?")==0      ||
		strcmp(argv[1],"-h")==0     ||
		strcmp(argv[1],"--help")==0 ||
		strcmp(argv[1],"help")==0 ||
		strcmp(argv[1],"/h")==0     ||
		strcmp(argv[1],"/?")==0 ))
   {
      printargs(args);
      osEnd();
      exit(OS_EXIT_OK);
   }

   if(!parseargs(args,argc,argv))
   {
      osEnd();
      exit(OS_EXIT_ERROR);
   }

   jbNewList(&list);

   /* get outbound dir */

   if((var=getenv("CMOUTBOUND")))
      cfg_Dir=var;

   else
      cfg_Dir=OS_CURRENT_DIR;

   if(args[ARG_DIRECTORY].data)
      cfg_Dir=(uchar *)args[ARG_DIRECTORY].data;

   /* Get zone */

   if((var=getenv("CMOUTBOUNDZONE")))
      cfg_Zone=atoi(var);

   else
      cfg_Zone=2;

   if(args[ARG_ZONE].data)
      cfg_Zone=atoi((uchar *)args[ARG_ZONE].data);

   /* Get pattern */

   strcpy(cfg_Pattern.Zone,"*");
   strcpy(cfg_Pattern.Net,"*");
   strcpy(cfg_Pattern.Node,"*");
   strcpy(cfg_Pattern.Point,"*");

   if(args[ARG_PATTERN].data)
   {
      if(!Parse4DPat((uchar *)args[ARG_PATTERN].data,&cfg_Pattern))
      {
         printf("Invalid node pattern \"%s\"\n",(uchar *)args[ARG_PATTERN].data);
         osEnd();
         exit(OS_EXIT_ERROR);
      }
   }

   /* Get verbose flag */

   cfg_Verbose=FALSE;

   if(args[ARG_VERBOSE].data)
      cfg_Verbose=TRUE;

   /* Real program starts here */

   printf("CrashListOut " VERSION "\n\n");

   scandir_dir = NULL;
   scandir_boss = NULL;

   if(!osScanDir(cfg_Dir,scandirfunc))
   {
		ulong err=osError();
      printf("Failed to scan directory %s\n",cfg_Dir);
		printf("Error: %s",osErrorMsg(err));		
      return(FALSE);
   }

   sortlist(&list);

   printf("%-8.8s %-17.17s %-16.16s %7.7s   %s\n\n",
      "Type","Node","Last Change","B/F","File");

   if(list.First)
   {
      for(fe=(struct fileentry *)list.First;fe;fe=fe->Next)
      {
         if(fe->type == TYPE_REQUEST) DisplayReq(fe);
         else if(fe->flow)            DisplayFlow(fe);
         else                         DisplayPkt(fe);
      }

      if(!cfg_Verbose) printf("\n");
      printf("Totally %s bytes in %lu files to send, %lu requests.\n",unit(TotalBytes),TotalFiles,TotalRequests);
   }
   else
   {
      printf("Outbound directory is empty.\n");
   }

   jbFreeList(&list);

   exit(OS_EXIT_OK);
}


